docs/proposals/20230117-cluster-level-identity.md

```diff
@@ -0,0 +1,166 @@
+   this capability
+## Non-goals
+
+1. This proposal does not solve multitenancy across a single cluster, for example, control plane in  Tenant A worker
+  nodes in Tenant B.
+
+## Proposal
+
+### User Stories
+
+#### Story 1 - Use a Cluster specific OCI user principal
+
+A large organization typically consists of many smaller OCI tenancies and compartments. Such a large organization
+may run a single Cluster API management cluster in a managed service model by a single team, let us call as Cluster 
+Ops team. The Cluster Ops team will have their own api or tools for customers to create and operate workload clusters.
+Cluster Ops team will want to make sure that individual teams use OCI user principal specific to their organization to 
+create the workload clusters so that the OCI resources are created in the correct tenancy/compartment and individual
+team are not able to create clusters in tenancies or compartments in which they do not have access. In order to achieve
+this goal, the Cluster Ops team will not use a single user principal to create all workload clusters, instead, the 
+individual workload clusters has to be created using OCI user principals tied to individual workload clusters.
+
+### Custom Resource Changes
+
+The following fields will be added to OCIClusterSpec custom resource.
+
+```go
+// OCIClusterSpec defines the desired state of OciCluster
+type OCIClusterSpec struct {
+  // IdentityRef is a reference to an identity(principal) to be used when reconciling this cluster
+  // +optional
+  IdentityRef *corev1.ObjectReference `json:"identityRef,omitempty"`
+}
+```
+The following custom resources will be added.
+
+```go
+
+type PrincipalType string
+
+const (
+    // UserPrincipal represents a user princpal.
+    UserPrincipal PrincipalType = "UserPrincipal"
+)
+
+// OCIClusterIdentitySpec defines the parameters that are used to create an OCIIdentity.
+type OCIClusterIdentitySpec struct {
+  // Type is the type of OCI Principal used.
+  // UserPrincipal is the only supported value
+  Type PrincipalType `json:"type"`
+  
+  // PrincipalSecret is a secret reference which contains the authentication credentials for the principal.
+  // +optional
+  PrincipalSecret corev1.SecretReference `json:"clientSecret,omitempty"`
+
+  // AllowedNamespaces is used to identify the namespaces the clusters are allowed to use the identity from.
+  // Namespaces can be selected either using an array of namespaces or with label selector.
+  // An empty allowedNamespaces object indicates that OCIClusters can use this identity from any namespace.
+  // If this object is nil, no namespaces will be allowed (default behaviour, if this field is not provided)
+  // A namespace should be either in the NamespaceList or match with Selector to use the identity.
+  //
+  // +optional
+  // +nullable
+  AllowedNamespaces *AllowedNamespaces `json:"allowedNamespaces"`
```

Comment:
The desired behaviour is not clear here. What does such an enforcement achieve ?
Additionally , shouldn't default behaviour be namespace "default", rather than not allowed.

Reply:
I had to use the behavios as both Azure and AWS uses thie behvaiour for cluster idenityt and if we change it, it will be confusing for the users, please see https://cluster-api-aws.sigs.k8s.io/topics/multitenancy.html and https://capz.sigs.k8s.io/topics/multitenancy.html#allowednamespaces

Reply:
ok, thanks.